### ISSUE TITLE
`DateTimeType::marshal()` occasionally returns string instead of object

### ISSUE THREAD

**Issue by chinpei215** [cakephp/cakephp#10025]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3.11

### What you did

Create a form containing a text input field for a datetime property.
```php
echo $this->Form->create($article);
echo $this->Form->input('published', ['type' => 'text']);
echo $this->Form->button('Submit');
echo $this->Form->end();
```

Dump the datetime property after `patchEntity()`.
```php
$article = $this->Articles->get(1);
$this->Articles->patchEntity($article, $this->request->data);
if ($this->Articles->save($article)) {
    var_dump($article->published);
}
$this->set('article', $article);
```

### What happened

When the posted value is `2017-01-01 00:00:00`, I get:
```
object(Cake\I18n\FrozenTime)#156 (3) {
  ["time"]=>
  string(25) "2017-01-01T00:00:00+09:00"
  ["timezone"]=>
  string(10) "Asia/Tokyo"
  ["fixedNowTime"]=>
  bool(false)
}
```

However, when the posted value is `2017-01-01 00:00`, I get:
```
string(16) "2017-01-01 00:00"
```

### What you expected to happen

I always get:
```
object(Cake\I18n\FrozenTime)#156 (3) {
  ["time"]=>
  string(25) "2017-01-01T00:00:00+09:00"
  ["timezone"]=>
  string(10) "Asia/Tokyo"
  ["fixedNowTime"]=>
  bool(false)
}
```

### Related code
When the input format is not `Y-m-d H:i:s`, `DateTimeType::marshal()` returns a string.
https://github.com/cakephp/cakephp/blob/3b341696e13ad6aed51e330d8e54479a41780512/src/Database/Type/DateTimeType.php#L164-L166

The following line would also return a string.
https://github.com/cakephp/cakephp/blob/3b341696e13ad6aed51e330d8e54479a41780512/src/Database/Type/DateTimeType.php#L170-L172

**Comment by lorenzo** [comment]
But `"2017-01-01 00:00"` is not a valid datetime format, it is only valid for dates. I'm more inclined to say that the validation was too relaxed and that the marshalled should not have put a value in the entity if it could not be marshalled correctly

**Comment by chinpei215** [comment]
I am not sure what kind of format is valid. `2017/01/01 00:00:00` would also cause the same issue.

Also, a value of `<input type="datetime-local">` would be `2017-01-01T00:00:00`. (Currently, `Validation::datetime()` doesn't allow this format though).

**Comment by lorenzo** [comment]
Are you using the `useLocaleParser()` feature?

**Comment by chinpei215** [comment]
No, I am. This code is not executed if `_useLocaleParser` is true.
https://github.com/cakephp/cakephp/blob/3b341696e13ad6aed51e330d8e54479a41780512/src/Database/Type/DateTimeType.php#L160-L166

**Comment by ADmad** [comment]
With local parsing turned off `Y-m-d H:i:s` would be the only acceptable format. So the marshaller return back string for all other formats is expected isn't it?

**Comment by chinpei215** [comment]
> So the marshaller returning back string for all other formats is expected isn't it?

@admad The `@return` annotation is `\Cake\I18n\Time|\DateTime`. So no one would expect it returns strings. I was surprised when I got a fatal error  from the following code:
```php
if ($this->Articles->save($article)) {
    $article->published->format('Y-m-d H:i:s');
}
```
If it is a feature, I would hesitate to use the `DateTimeType` class.

What is the benefit of returning strings? I cannot see any good reasons.

**Comment by markstory** [comment]
Shouldn't the marshaller be able to marshal values that pass validation? Using text input and no validation means you always run the chance of getting text out of the marshaller.

**Comment by chinpei215** [comment]
I think we should remove [line 164~166](https://github.com/cakephp/cakephp/blob/3b341696e13ad6aed51e330d8e54479a41780512/src/Database/Type/DateTimeType.php#L164-L166). I think we can assume that the value is already validated, because `Time` or `FrozenTime` object can be created from the string. Although it would allow to convert `+1 day` to `2017-01-15` for example, it is a kind of thing. If users don't want it, they can validate it with the `datetime` validation rule.

Regarding [line 171](https://github.com/cakephp/cakephp/blob/3b341696e13ad6aed51e330d8e54479a41780512/src/Database/Type/DateTimeType.php#L171), I think we should return `null` instead of `$value`. Currently, it may return `0xFF` or `0.1` or something.

**Comment by markstory** [comment]
I think those lines are to help prevent ambiguous input that result in valid but unintended date values from being marshalled.

**Comment by chinpei215** [comment]
> ambiguous input

Do you mean `01/01/01 00:00:00` for example? But currently, the returned value will be embedded into SQL statement without using timezone settings defined in `app.php`. I don't think it is a correct behavior:
```sql
SET published = '01/01/01 00:00:00'
```
If we could return `Time` or `FrozenTime` object, it could use the timezone settings:
```sql
SET published = '2001-01-01 09:00:00'
```

**Comment by chinpei215** [comment]
I forgot about  #6571 . My example above is not true. So is it a feature? Then I will open a pull request to change the return type to `\Cake\Chronos\ChronosInterface|\DateTimeInterface|string|null`.

**Comment by markstory** [comment]
Related to #10587

**Comment by markstory** [comment]
@chinpei215 Do you think this can be closed when #10587 merges? Or should we do more work on this?

**Comment by chinpei215** [comment]
@markstory Probably, we may have to change the return annotation of `marshal()`. Now it says:
```php
@return \Cake\I18n\Time|\DateTime
```
But actually, it returns string in some cases.

**Comment by chtheis** [comment]
I ran into the same problem. What makes things worse that there is no way (known to me) to detect if marshalling was successful or not. If not the input value is passed as is to the database (MariaDB in my case) and if you are lucky the database throws an error. In which case the user would know something is wrong, but no details are given. And if you are not so lucky the database would write any value (happens to me on one PC).

If the marshaller shall check for ambiguous input it should not do so with one format only, "yyyy-mm-dd": there are hardly any cultures in which this is a common way to write dates. Similar for times, half of the world is using the am/pm system which is OK for validation but not for the marshaller. So either the check shall be done in a different way (can we compare the timestamp of the DateTime object against PHP strtotime?), or it should try more patterns. 

It would help me a lot if there were a way I could set an array of format strings for that crosscheck.

**Comment by lorenzo** [comment]
The way to detect errors is to have validation, using validatora or application rules. 

I agree, though, that there should be a way to know whether or not the type casting was successful.

**Comment by markstory** [comment]
It enables datetime values to be parsed based on the current user's locale using the date format rules defined in the intl extension. These are the same rules/formats that CakePHP uses to convert datetime 
objects into strings as well.

**Comment by chtheis** [comment]
I tried and it didn't help me, and I think it couldn't: it will try to use one pattern in one locale to parse the input. At least I guess so because parsing is done in an internal PHP function. But there is no way I could predict which way a user will write the input. Even a user with language set to English could still be a German expatriate. I could even get a wrong date: with my browser set to English "14.7.2018" is returned as 7th Feb 2019.

**Comment by chtheis** [comment]
That's not the complete truth :)
PHP's strtotime and other functions do a good job figuring out which format may be used, and, by default, the CakePHP's marshallers for date and time use that. The problem is, that then the result is converted back to a string and compared with the input and that will force that only one format is accepted, though the _compare method even allows for multiple format strings used one after another. If we had the most common formats there everything would be fine.

**Comment by ADmad** [comment]
> PHP's strtotime and other functions do a good job figuring out which format may be used..

Nope they are not. Given a string `01/07/2018` `strtotime()` and `DateTime::__construct()` will assume it's in `m/d/y` format which would be wrong if you are not in USA.

That is why we provide the option of locale based date string parsing which uses `IntlDateFormatter` which requires the format to be specified or picks one based on the locale set.

**Comment by ADmad** [comment]
> Going back to the original problem of returning strings when datetime objects are expected. @chinpei215's suggestion of returning null when we currently return a string looks like a viable approach on the surface.

But wouldn't that mean the text field won't be populated with user entered value when the form is rendered again with validation errors?

**Comment by markstory** [comment]
@ADmad Yes.

**Comment by markstory** [comment]
Sounds like people agree that returning `null` instead of strings will be an improvement. I can put that in place for 3.7

**Comment by chtheis** [comment]
Will returning null make things really better? Will null not be passed s NULL to the database?
If the column accepts NULL then the database will silently write something wrong. And if not you'll see an error about NULL not being allowed and one wonders where that value is coming from. I don't think many people are aware of the behaviour of the date / time marshallers.
If the input cannot be parsed I would prefer that saving the entity fails with an error for the resp. field.  E.g. throwing an exception which is caught by the caller. 
But that may be a breaking change. Though, in that case, something is broken anyway and the change an improvement.

And to my other point: because input data format has to be fixed (you convinced me here) then the date and time validator should have an option to accept that format only. I can achieve that with a custom validator, but does it make sense that everyone has to write that themselves? and the regexp taking care of months with 28-31 days is not straightforward for someone not familiar with that. And finally I'm not sure if many users of CakePHP are aware of that. At least I was happily using the validator and then had to spent quite some time to figure out what is happening.

**Comment by ADmad** [comment]
> If the column accepts NULL then the database will silently write something wrong.
> If the input cannot be parsed I would prefer that saving the entity fails with an error for the resp. field. E.g. throwing an exception which is caught by the caller.

I agree. The db type classes needs some type of error reporting when they fail to reliably marshal a value, instead of current behavior of just returning the input value or `null`.

> then the date and time validator should have an option to accept that format only

There are simply too many possible formats and it's not practically possible to provide validators for all. This is why when marshalling the datetime object is again converted to string and compared to the initial string value to ensure the marshalled object representation is correct. So in case of datetime string validation is just a first line of defence and not a foolproof check.

**Comment by markstory** [comment]
We've been trying to avoid raising errors from the marshaller because it makes presenting validation errors more complicated. If the marshaller returned null for incorrect datetime strings, then validiation could at least be used to mark the field as required, or as having the incorrect format.

> And to my other point: because input data format has to be fixed (you convinced me here) then the date and time validator should have an option to accept that format only. 

@chtheis It does. The [date and datetime](https://api.cakephp.org/3.6/class-Cake.Validation.Validation.html#_date) validators let you control the accepted formats.

**Comment by chtheis** [comment]
Sorry for the late answer but I'm away from home and can't read mails every day.

@ADmad: DateType formats the result of conversion with "Y-m-d" and compares the outcome with the input, so we need a validator for only that one. But the current validator accepts may other formats as well which would not pass DateType.

@markstory: Date validator can be restricted to 'ymd', but that fixes the order of year, month, and day only. According to documentation it still allows for different separators, for years without century and for single digit days and months. Which is still to broad to reject what would not pass DateType marshaller. For everything else you need to write your own pattern.

I may have overseen something, so please be patient if I did.

**Comment by markstory** [comment]
This will be fixed in 4.0.0
